Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -489,93 +504,133 @@ void CGContextAddRects(CGContextRef context, const CGRect* rects, unsigned count
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+
+    CGPathMoveToPoint(context->Path(), &context->CurrentGState().transform, x, y);
 }
 
 /**
  @Status Interoperable
 */
 void CGContextAddArc(CGContextRef context, CGFloat x, CGFloat y, CGFloat radius, CGFloat startAngle, CGFloat endAngle, int clockwise) {
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+
+    CGPathAddArc(context->Path(), &context->CurrentGState().transform, x, y, radius, startAngle, endAngle, clockwise);
 }
 
 /**
  @Status Interoperable
 */
 void CGContextAddArcToPoint(CGContextRef context, CGFloat x1, CGFloat y1, CGFloat x2, CGFloat y2, CGFloat radius) {
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+
+    CGPathAddArcToPoint(context->Path(), &context->CurrentGState().transform, x1, y1, x2, y2, radius);
 }
 
 /**
  @Status Interoperable
 */
 void CGContextAddEllipseInRect(CGContextRef context, CGRect rect) {
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+
+    CGPathAddEllipseInRect(context->Path(), &context->CurrentGState().transform, rect);
 }
 
 /**
  @Status Interoperable
 */
 void CGContextAddPath(CGContextRef context, CGPathRef path) {
     NOISY_RETURN_IF_NULL(context);
-    // The Apple SDK docs imply that passing a NULL path is valid.
-    // So avoid calling into the backing if NULL.
-    if (path) {
-        UNIMPLEMENTED();
+    if (!path) {
+        return;
+    }
+
+    if (!context->HasPath()) {
+        // If we don't curerntly have a path, take this one in as our own.
+        woc::unique_cf<CGMutablePathRef>& currentPathRef{context->Impl().currentPath};
+        currentPathRef.reset(CGPathCreateMutableCopy(path));
+        return;
     }
+
+    CGPathAddPath(context->Path(), &context->CurrentGState().transform, path);
 }
 
 /**
  @Status Stub
 */
 void CGContextReplacePathWithStrokedPath(CGContextRef context) {
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+
+    if (!context->HasPath()) {
+        return;
+    }
+
+    // We're avoiding .Path() here since we want the actual unique_cf<CGMutablePathRef>.
+    woc::unique_cf<CGMutablePathRef>& currentPathRef{context->Impl().currentPath};
+
+    auto& state = context->CurrentGState();
+
+    // TODO GH#xxxx When CGPathCreateCopyByStrokingPath is no longer stubbed, remove the diagnostic suppression.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+
+    woc::unique_cf<CGPathRef> newPath{CGPathCreateCopyByStrokingPath(currentPathRef.get(), &state.transform, state.lineWidth, (CGLineCap)state.strokeProperties.startCap, (CGLineJoin)state.strokeProperties.lineJoin, state.strokeProperties.miterLimit)};
+
+#pragma clang diagnostic pop
+
+    currentPathRef.reset(CGPathCreateMutableCopy(newPath.get()));
 }
 
 /**
  @Status Interoperable
 */
 bool CGContextIsPathEmpty(CGContextRef context) {
     NOISY_RETURN_IF_NULL(context, StubReturn());
-    UNIMPLEMENTED();
-    return StubReturn();
+
+    if (!context->HasPath()) {
```

Comment:
ughghg i got this backwards because i suck

---
In reply to: [85366792](https://github.com/Microsoft/WinObjC/pull/1240#discussion_r85366792) [](ancestors = 85366792)